### PR TITLE
adding implementation for drawing circles.

### PR DESCRIPTION
**Pull request by Gr1mR3aver**
## Description

Adding implementation of drawing filled and empty circles

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by Gr1mR3aver** [comment]
any idea why the keyboard linting fails for this one?

**Comment by tzarc** [comment]
You can ignore it -- linting is currently still being written on main QMK, the painter branch is out of date.

File: drivers/painter/fallback/qp_fallback.c [hunk 1]
```diff
@@ -95,3 +95,137 @@ bool qp_fallback_rect(painter_device_t device, uint16_t left, uint16_t top, uint
...
+    qp_fallback_circle_drawpixels(device, x, y, xcalc, ycalc, hue, sat, val, filled);
+    while (xcalc < ycalc) {
+        xcalc++;
+        if (err < 0) {
+            err += (xcalc << 1) + 1;
+        } else {
+            ycalc--;
+            err += ((xcalc - ycalc) << 1) + 1;
+        }
+        qp_fallback_circle_drawpixels(device, x, y, xcalc, ycalc, hue, sat, val, filled);
```
Comment (tzarc) [inline drivers/painter/fallback/qp_fallback.c:117]: This file will need a declaration for this function above this line -- it's lower in the file so isn't yet visible.

File: drivers/painter/fallback/qp_fallback.c [hunk 2]
```diff
@@ -95,3 +95,137 @@ bool qp_fallback_rect(painter_device_t device, uint16_t left, uint16_t top, uint
...
+    For filled circles, we can draw horizontal lines between each pair of
+    pixels with the same final value of y.
+
+    Two special cases exist and have been optimized:
+    1) offsetx == offsety (the final point), makes half the coordinates
+    equivalent, so we can omit them (and the corresponding fill lines)
+    2) offsetx == 0 (the starting point) means that some horizontal lines
+    would be a single pixel in length, so we write individual pixels instead.
+    This also makes half the symmetrical points identical to their twins,
+    so we only need four points or tqo points and one line
```
Comment (tzarc) [inline drivers/painter/fallback/qp_fallback.c:139]:
```suggestion
    so we only need four points or two points and one line.
```

**Review by tzarc — CHANGES_REQUESTED** [review]
This will also need a `qmk cformat` and some plumbing to hook up a `qp_circle` API, but I can handle those after merge. Thanks!

File: drivers/painter/fallback/qp_fallback.c (hunk 1 above)
Comment (Gr1mR3aver) [reply drivers/painter/fallback/qp_fallback.c:120]: added

File: drivers/painter/fallback/qp_fallback.c (hunk 2 above)
Comment (Gr1mR3aver) [reply drivers/painter/fallback/qp_fallback.c]: fixed

**Comment by tzarc** [comment]
Haven't forgotten about these, been super busy getting ChibiOS upgraded in main QMK. That's done, so I should be able to get around to these in the next couple of days.

**Comment by tzarc** [comment]
Unsure this is working correctly, I get diamonds when drawing circles:

image

**Comment by tzarc** [comment]
And 20px x 12px ellipses:

image

**Comment by tzarc** [comment]
Decided to properly look at the line drawing from before, too -- it's not functioning correctly either.
The issue is that all integers are unsigned, rather than signed, and the algorithms seem to be non-functional otherwise.
Will have a look and see if there's some modifications that can be made in order to get things functional.

**Comment by tzarc** [comment]
Worked out a fix, will apply separately.

**Comment by tzarc** [comment]
After swapping to signed arithmetic:

image

**Comment by Gr1mR3aver** [comment]
Yeah, they need to be unsigned integers, or the algorithms don't work.  Why is unsigned an issue?

Sent from my T-Mobile 4G LTE Device
Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Gr1mR3aver** [comment]
apologies for the mistake.  went back and looked at my testing, and I see my error.

**Comment by tzarc** [comment]
All good mate, it's in, and it's working! Thanks again :)

**Pull request merged**